### PR TITLE
Logger

### PR DESCRIPTION
**Pull request by DannyPeck**
Redesigned Conversion::Variant and Conversion::VariantVisitor into a Result class with a Variant typedef inside (Result::Variant) and an internal protected VariantVisitor class.  Finished writing Unit Tests for the Logger system and all of its subsystems.

### CODE REVIEW

**Comment by technetos** [comment]
in ```ConvertableTo``` the ```convert``` method is overkill and is implementing something that is inherently already part of c++.  

```convert``` should be replace with the following conversion function:

```virtual const operator ResultType (void) const = 0;```

The above declares a conversion function for the implicit conversion from what ever implements the ```ConvertableTo``` interface to what ever type ```ResultType``` is

Sources:
[http://en.cppreference.com/w/cpp/language/cast_operator#Explanation](http://en.cppreference.com/w/cpp/language/cast_operator#Explanation)

**Comment by technetos** [comment]
The variant isn't necessary.  The compiler is going to try exact type candidates first, then it will try implicit conversion candidates if no exact candidates are found.  **We don't need infrastructure to handle an either-or scenario, C++ does this already.**

**Review by technetos — CHANGES_REQUESTED** [review]
See my comments on the PR

**Review by technetos — APPROVED** [review]

File: C++/Logger/makefile [hunk 1]
```diff
@@ -13,57 +13,60 @@ CONVERSION = conversion
 TESTS = tests
 TEST_TYPES = test_types
 
+# compiler
+CC = g++
```
Comment (xmclark) [inline C++/Logger/makefile]: Thank you for allowing the compiler to be easily configurable in make.

File: C++/Logger/TypeIs.h [hunk 2]
```diff
@@ -2,19 +2,19 @@
 #define _TYPE_IS_H_
 
 template <typename T>
-struct type_is
+struct TypeIs
```
Comment (xmclark) [inline C++/Logger/TypeIs.h]: Is this our preferred naming convention for template metafunctions? The old way was more akin to the STL style which I prefer, but the PascalCase is fine. 

EDIT: I see in the previous comit that you mention this is the preferred style for the project.

**Review by xmclark — COMMENTED** [review]
I'm cool with this.

File: C++/Logger/TypeIs.h (hunk 2 above)
Comment (DannyPeck) [reply C++/Logger/TypeIs.h]: I know we stated the preferred style for the project but I am inclined to agree with you for template metafunctions.

**Comment by technetos** [comment]
Approved!

File: C++/Logger/TypeIs.h (hunk 2 above)
Comment (technetos) [reply C++/Logger/TypeIs.h]: I too prefer `type_is` code updated.

Comment (technetos) [inline C++/Logger/TypeIs.h]: Also how do you want to name the files themselves? `TypeIs.h` or `type_is.h`

**Pull request merged**